File: contest_2/tasks/H.py
import random

class Node:
    def __init__(self, value):
        self.value = value
        self.priority = random.randint(1, 10**9)
        self.size = 1
        self.left = None
        self.right = None

def update_size(node):
    if node:
        node.size = 1 + get_size(node.left) + get_size(node.right)

def merge(left, right):
    if not left or not right:
        return left or right

    if left.priority > right.priority:
        left.right = merge(left.right, right)
        update_size(left)
        return left
    else:
        right.left = merge(left, right.left)
        update_size(right)
        return right

def split(node, key):
    if not node:
        return None, None

    if get_size(node.left) >= key:
        left, node.left = split(node.left, key)
        update_size(node)
        return left, node
    else:
        node.right, right = split(node.right, key - get_size(node.left) - 1)
        update_size(node)
        return node, right

def insert(node, value, pos):
    left, right = split(node, pos)
    return merge(merge(left, value), right)

def get_size(node):
    return node.size if node else 0

def print_tree(node):
    if node:
        print_tree(node.left)
        print(node.value, end=' ')
        print_tree(node.right)

n, m = map(int, input().split())
tree = None

for i in range(n):
    tree = insert(tree, Node(i+1), i)

for _ in range(m):
    l, r = map(int, input().split())
    tree_left, tree = split(tree, l-1)
    tree_mid, tree_right = split(tree, r-l+1)
    tree = merge(tree_mid, merge(tree_left, tree_right))

print_tree(tree)

File: contest_2/tasks/H.cpp
#include <random>
#include <fstream>
#include <iostream>

class Node {

public:
    int key;
    int priority;
    int size = 1;
    Node* left = nullptr;
    Node* right = nullptr;

    Node(int key) : key(key), priority(rand()) {}
}; 

class Treap {

public:
    Node* root = nullptr;

    static int getSize(Node* node) {
        if (node == nullptr) {
            return 0;
        }
        return node->size;
    }

    static void update(Node*& node) {
        if (node != nullptr) {
            node->size = getSize(node->left) + 1 + getSize(node->right);
        }
    }

    static Node* merge(Node* a, Node* b) {
        if (a == nullptr || b == nullptr) {
            if (a != nullptr) {
                return a;
            }
            return b;
        }

        if (a->priority > b->priority) {
            a->right = merge(a->right, b);
            update(a);
            return a;
        } else {
            b->left = merge(a, b->left);
            update(b);
            return b;
        }
    }

    static void split(Node* node, int k, Node*& a, Node*& b) {
        if (node == nullptr) {
            a = nullptr;
            b = nullptr;
            return;
        }

        if (getSize(node->left) < k) {
            split(node->right, k - getSize(node->left) - 1, node->right, b);
            a = node;
        } else {
            split(node->left, k, a, node->left);
            b = node;
        }

        update(a);
        update(b);
    }

    void add(int key) {
        root = merge(root, new Node(key));
    }

    void insert(int i, int x) {
        Node* less;
        Node* greater;
        split(root, i, less, greater);
        root = merge(merge(less, new Node(x)), greater);
    }

    int get(int i) {
        Node* less;
        Node* equal;
        Node* greater;

        split(root, i, less, greater);
        split(greater, 1, equal, greater);
        int value = equal->key;
        root = merge(merge(less, equal), greater);
        return value;
    }

    void remove(int index) {
        Node* less;
        Node* equal;
        Node* greater;
        split(root, index, less, greater);
        split(greater, 1, equal, greater);
        root = merge(less, greater);
    }

    void inorderTraversal(Node* node, std::vector<int>& arr) {
        if (node == nullptr) {
            return;
        }
        inorderTraversal(node->left, arr);
        arr.push_back(node->key);
        inorderTraversal(node->right, arr);
    }

    void printArray() {
        std::vector<int> arr;
        inorderTraversal(root, arr);
        std::cout << arr.size() << std::endl;
        for (int i = 0; i < arr.size(); i++) {
            std::cout << arr[i] << " ";
        }
        std::cout << std::endl;
    }

    int size() {
        return getSize(root);
    }

    void shift(int l, int r) {
        Node* less;
        Node* equal;
        Node* greater;

        split(root, l, less, greater);
        split(greater, r - l + 1, equal, greater);
        root = merge(merge(equal, less), greater); 
    }
};

int main() {
    Treap treap;

    int n;
    int m;

    std::cin >> n >> m;

    for (int i = 0; i < n; i++) {
        treap.add(i + 1);
    }

    for (int i = 0; i < m; i++) {
        int l;
        int r;

        std::cin >> l >> r;
        treap.shift(l - 1, r - 1);
    }

    for (int i = 0; i < treap.size(); i++) {
        std::cout << treap.get(i) << " ";
    }

    return 0;
}
